docs-src/concepts.md

```diff
@@ -112,42 +114,123 @@ TODO
+`Gateways` that define Internet-facing vs private, internal applications.
 
 ```yaml
-apiVersion: networking.acme.io/v1alpha1
-kind: CloudLB
+kind: GatewayClass
+metadata:
+  name: internet
+  ...
+---
+kind: GatewayClass
 metadata:
-  name: from-internet
-networkTier: auto
-ipAllocation: auto
+  name: private
+  ...
 ```
 
-We expect that one or more GatewayClasses will be created by the infrastructure
-provider for the user as part of their cloud infrastructure. If the cluster
-operator wishes to customize their `GatewayClasses`, they are free to do so,
-interacting with the custom resource parameters. Example parameters
-`GatewayClass.controller` resource will include cluster-level defaults and
-constraints.
+The user of the classes will not need to know *how* `internet` and `private` are
+implemented, whether they are cluster-based or cloud-based LBs.
 
 > The community may converge around a common set of GatewayClass names with
 > well-known meanings (e.g. `internet`, `private`) to ease interoperability
 > between platforms.
 
+#### GatewayClass parameters
+
+Providers of the `Gateway API` may need to pass parameters to their controller
+as part of the class definition. This done using the
+`GatewayClass.spec.parametersRef` field:
+
+```yaml
+# GatewayClass for Gateways that define Internet-facing applications.
+kind: GatewayClass
+metadata:
+  name: internet
+spec:
+  controller: "acme.io/gateway-controller"
+  parametersRef:
+    apiVersion: core/v1
+    kind: ConfigMap
+    namespace: acme-system
+    name: internet-gateway
+---
+kind: ConfigMap
+metadata:
+  name: internet-gateway
+  namespace: acme-system
+data:
+  ip-address-pool: internet-vips
+  ...
+```
+
+The type of object referenced by `GatewayClass.spec.parametersRef` will depend on
+the provider itself. A `core.ConfigMap` is used in the example above, but some
+controllers may opt to use a `CustomResource` for better schema validation.
+
+#### GatewayClass status
+
+`GatewayClasses` need to be validated by the provider to ensure that the
+configured parameters are valid. The validity of the class will be signaled to
+the user via `GatewayClass.status`:
+
+```yaml
+kind: GatewayClass
+...
+status:
+  conditions:
+  - type: InvalidParameters
+    ...
+```
+
+A new `GatewayClass` will start with the `InvalidParameters` condition set to
+`Unknown`. At this point the controller has not seen the configuration. Once the
+controller has processed the configuration, the condition will be set to
+`False`.
+
+```yaml
+kind: GatewayClass
+...
+status:
+  conditions:
+  - type: InvalidParameters
+    status: False
+    ...
+```
+
+If there is an error in the `GatewayClass.spec`, the conditions will be
+non-empty and contain information about the error. 
+
+```yaml
+kind: GatewayClass
+...
+status:
+  conditions:
+  - type: InvalidParameters
+    status: True
+    Reason: InvalidIPAddressPool
+    Message: The ip-address-pool my-ip-addresses does not exist
+```
+
 ### Gateway
```

Comment:
How do you glue in SNI and DNS?

Another point I want to make is in our current setup we run 2 layer loadbalancer with one definition.
An ingress creates an ALB, via kube-ingress-aws-controller, which also writes the status field to point to the provisioned ALB. The other ingress controller, skipper, does http routing for the ingress definition. Skipper is in the TargetGroup of the ALB, so gets all traffic. I guess a gatewayclass=alb would be created. If I create a gateway with class=alb, does it provision an instance of the kube-ingress-aws-controller or does it create the ALB?
Later in this file it states gateway is a 1:1 mapping with LB, which would mean we would need to provision an ALB.
Right now we are already much better, than the idea of ingress v2, because we create ALBs based on ingress objects. For example if we have >25 host definitions, that are not covered by wildcard certificates we *automatically* provision ALBs on behalf of the user. The user does not have to care at all about ALB!
If every user creates its own gateway, then we just waste money. If we pre-provision one gateway, it does not do automatically the ALB split as I wrote for the case when we need >25 certificates, or an internal ALB for another example.

Reply:
Adding to @szuecs point, TLS configurations could face similar challenges.

In the case of cloud-provider issued and managed SSL certificates, they are provisioned and attached to loadbalancers for terminating the secure connection within the boundaries of the infrastructure provider. In that particular case, the "Cluster operator"-persona can't possibly specify TLS secrets on the Gateway resource, and even less so on the GatewayClass resource configuration.

Reply:
SNI and DNS aren't part of GatewayClass level at all.

GatewayClass should be viewed (for those familiar w/ Ingress) as the same as IngressClass, with the additional flexibility of actually provisioning names/resource representation on the cluster and attachment of configuration.

The API does not (and should make this clear with working) state how Gateway, GatewayClass reflects into "implementation related provisioning", e.g. if your implementation multiplexes LBs to meet the requirements of the API, you may end up having a different lifecycle for the implementation bits. Just like how a serverless API does not talk about specific VMs/containers etc, but from a user's standpoint there is a lifecycle associated with the event-triggered function that is run in a serverless platform.

From a user's standpoint, the Gateway is an entity that triggers a configuration change to provision an "LB". I use quotes here because it may be a conceptual thing stitched together from many physical pieces w/ their own lifecycles etc. For the most part, however, we expect that 1-1 relationship is meaningful : cloud implementations will provision/delete resources on Gateway creation/deletion; proxy providers will modify config on Gateway create/deletion etc.

Reply:
@bowei that does not clarify my concerns.
We run 150 clusters and I will likely not manage in all clusters the (for us useless) gateway resource. Now if a user creates a gateway and later tears down it, because he does not use it, it impacts all other users that depend on it. On the other hand we could also provision for each application a gateway, which means we have now thousands of additional load balancers provisioned and the only one who is happy is AWS, because of the bill increase.

Reply:
This depends on the behavior of your controller whether or not cloud resources correspond to a cloud load balancer.
From your description, it sounds like your deployment is one cloud LB => proxy software that you run => services. 

This sounds similar to how ingress-nginx currently works:

All Ingress resources are served by proxy instances.
There are no cloud resources allocated automatically -- this is done by the user, typically one cloud LB allocated to point to the proxies. 
Gateway would be similar.

Basically -- how the deployment maps depends on how your controller implements it on your cloud.

Can you point to which controller you use and how it manages cloud resources?

Reply:
From my latest reading of the spec, it seems like it breaks down like this:
- GatewayClass allows you to run multiple ways of ingress in the same cluster.
- Gateway describes how the traffic gets to the load balancer (whatever that is). In the ingress controller case, the important part is that Gateway will be the thing that holds what the Ingress.status.loadBalancer section does now - the externally reachable addresses for the traffic to go to. In the cloud load balancer case, the Gateway represents the load balancer instance. In both cases, the Gateway spec represents the desired state for the load balancer (what port to listen on, etc.). This is not how I thought they were originally, I thought that Gateway corresponded more closely with what Ingress currently is, which is not the case. The current Ingress use cases are split between Gateway and Route.
- Routes describe how the traffic should be routed once it gets to the Gateway.
- Services are the same as now.

@bowei, how did I do? Is that a better description?

Reply:
That sounds correct. Although for in cluster proxy deployments, we typically see that the LoadBalancer routing traffic to the proxy is done by the user as something separate from the controller itself. 

The thing to remember regarding the underlying implementation resource model our goal is to support many different choices (e.g. merging config vs creating 1-1 representation w/ cloud LBs etc) because we know that Ingress implementations today have this kind of diversity. 

https://docs.google.com/drawings/d/1vIOSjWttZ54j_xWAnAC2PFFKRCXTeHIzJwV3uBvvCNE

Reply:
@youngnick I added my use cases already. I am aware of the gw + httproute split, but what is the actual value for me?
As far as I see you create 3 crds, and as cluster operator I have now to maintain 2  CRDs for 150 clusters and before it was completely self service.

Reply:
If the Ingress resource meets your use cases -- it will continue to be maintained and improved.

BTW -- The direction of Kubernetes APIs is to put new features as CRDs, so this will not be limited just to this or networking.

Reply:
Side note:

> The thing to remember regarding the underlying implementation resource model our goal is to support many different choices (e.g. merging config vs creating 1-1 representation w/ cloud LBs etc) because we know that Ingress implementations today have this kind of diversity.

We should consider adding this to the Spec itself to standardize the configuration across providers. We can make it optional for providers to implement .